// utils/utils.js
var fs = require('fs');
var babyparse = require('babyparse');
var present = require('present');

function readCSV(filename){
  return babyparse.parse(fs.readFileSync(filename, 'utf8'));
};

function getTime(){
  return present()
}

function isNumber(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
};
function closest(arr, closestTo){

    var closest = Math.max.apply(null, arr); //Get the highest number in arr in case it match nothing.
    for(var i = 0; i < arr.length; i++){ //Loop the array
        if(arr[i] >= closestTo && arr[i] < closest) closest = arr[i]; //Check if it's higher than your number, but lower than your closest value
    }
    return closest; // return the value
};

var erpWriter = function(erp, filename) {
 console.log('writing to csv')
 var supp = erp.support([]);
 var csvFile = fs.openSync(filename, 'w');
 fs.writeSync(csvFile,'Type,Parameter,Item,Value,Probability\n')
 supp.forEach(function(s) {supportWriter(s, Math.exp(erp.score([], s)), csvFile);})
 fs.closeSync(csvFile);
 console.log('writing complete.')
}

var supportWriter = function(s, p, handle) {
 var sLst = _.pairs(s);
 var l = sLst.length;

 for (var i = 0; i < l; i++) {
   fs.writeSync(handle, sLst[i].join(',')+','+p+'\n');
 }
}

function fillArray(value, len) {
  var arr = [];
  for (var i = 0; i < len; i++) {
    arr.push(value);
  }
  return arr;
};

function wpParseFloat(x){
  return parseFloat(x);
};

function containsObject(obj, list) {
    var i;
    for (i = 0; i < list.length; i++) {
        if (_.isEqual(list[i],obj)) {
            return true;
        }
    }

    return false;
}

module.exports = {
  readCSV: readCSV,
  isNumber: isNumber,
  fillArray: fillArray,
  erpWriter:erpWriter,
  closest:closest,
  getTime:getTime,
  wpParseFloat: wpParseFloat,
  containsObject:containsObject
};
